### PR TITLE
PLUGIN-719 : create temp table and load in batches of 10000 and then …

### PR DESCRIPTION
**Pull request by sahusanket**
…importing in actual table

### CODE REVIEW

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java [hunk 1]
```diff
@@ -392,26 +383,33 @@ private void importFromGcs(String projectId, TableReference tableRef, @Nullable
...
+      loadConfig.setDestinationTable(temporaryTableReference);
+      loadConfig.setWriteDisposition(JobInfo.WriteDisposition.WRITE_APPEND.toString()); // Always append to temp table
 
-      // Poll until job is complete.
-      waitForJobCompletion(bigQueryHelper.getRawBigquery(), projectId, jobReference);
+      Collection<List<String>> gcsPathsInBatches = splitListInBatches(gcsPaths, BQ_IMPORT_MAX_BATCH_SIZE);
+
+      int jobcount = 1;
+      for (List<String> gcsPathBatch :  gcsPathsInBatches) {
+        LOG.info(" Running for Batch {} with number of gcs paths : {}", jobcount, gcsPathBatch.size());
```
Comment (nitinmotgi) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java]: Not as info

File: src/test/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormatTest.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import org.apache.hadoop.mapreduce.JobContext;
+import org.junit.Assert;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+
```
Comment (nitinmotgi) [inline src/test/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormatTest.java:41]: Test is not actually testing the change

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java [hunk 3]
```diff
@@ -736,4 +771,9 @@ private static BigQuery getBigQuery(Configuration config) throws IOException {
       GCPUtils.loadServiceAccountCredentials(serviceAccount, isServiceAccountFile);
     return GCPUtils.getBigQuery(projectId, credentials);
   }
+  
+  protected static <T> Collection<List<T>> splitListInBatches(List<T> inputList, int batchSize) {
```
Comment (fernst) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java]: You can use  the `com.google.common.collect.Lists.partition` method to achieve this.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java (hunk 1 above)
Comment (fernst) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java]: +1

File: src/test/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormatTest.java (hunk 2 above)
Comment (sahusanket) [reply src/test/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormatTest.java:41]: Added some mocked unit tests to actually verify invocations.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java (hunk 1 above)
Comment (sahusanket) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java]: done

**Comment by fernst** [comment]
@sahusanket did you test this code with the BQ Sink? Specially when creating new tables or updating schema for an existing table.

**Comment by sahusanket** [comment]
@fernst  Yes I had tested the schema scenario. Please check https://docs.google.com/spreadsheets/d/1VQxtD3i6Gy4SONLypOeNRsCso7TzB39TgGJCDNp15S4 .I have tested these scenarios locally

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java (hunk 3 above)
Comment (sahusanket) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java]: used the suggested method. Thanks

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java [hunk 4]
```diff
@@ -421,6 +417,13 @@ private void importFromGcs(String projectId, TableReference tableRef, @Nullable
                                                         temporaryTableReference.getTableId(), table).execute();
       }
 
+      if (operation.equals(Operation.INSERT)) {
+        handleInsertOperation(tableRef, writeDisposition, loadConfig.getDestinationEncryptionConfiguration(), projectId,
+                              jobId, dataset, tableExists);
+      } else {
+        operationAction(tableRef, kmsKeyName, getJobIdForUpdateUpsert(conf), conf);
```
Comment (fernst) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java]: nitpick: Since this now covers the UPDATE/UPSERT operation, maybe this method should be renamed to handleUpdateUpsertOperation?

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java [hunk 5]
```diff
@@ -392,26 +381,33 @@ private void importFromGcs(String projectId, TableReference tableRef, @Nullable
...
-      // Insert and run job.
-      bigQueryHelper.insertJobOrFetchDuplicate(projectId, job);
+      loadConfig.setDestinationTable(temporaryTableReference);
+      loadConfig.setWriteDisposition(JobInfo.WriteDisposition.WRITE_APPEND.toString()); // Always append to temp table
 
-      // Poll until job is complete.
-      waitForJobCompletion(bigQueryHelper.getRawBigquery(), projectId, jobReference);
+      List<List<String>> gcsPathsInBatches = Lists.partition(gcsPaths, BQ_IMPORT_MAX_BATCH_SIZE);
+
+      int jobcount = 1;
```
Comment (fernst) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java:450]: add a comment here explaining what this logic does (load in batches of up to 10k files)

Comment (sahusanket) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java:450]: done

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java (hunk 4 above)
Comment (sahusanket) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java]: renamed it.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java [hunk 6]
```diff
@@ -267,12 +268,13 @@ private void importFromGcs(String projectId, TableReference tableRef, @Nullable
                                @Nullable String kmsKeyName, BigQueryFileFormat sourceFormat, String writeDisposition,
                                List<String> gcsPaths, PartitionType partitionType, @Nullable Range range,
                                @Nullable String partitionByField, boolean requirePartitionFilter,
-                               List<String> clusteringOrderList, boolean tableExists, String jobId)
-      throws IOException, InterruptedException {
+                               List<String> clusteringOrderList, boolean tableExists, Configuration conf)
+      throws Exception {
```
Comment (snehalakshmisha) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java]: Why are we throwing the generic exception instead of the specific ones?

**Review by snehalakshmisha — COMMENTED** [review]
Why are we throwing a generic exception instead of specific ones?

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java (hunk 6 above)
Comment (sahusanket) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java]: Because one of the old method was throwing "Exception". On including that in this method automatically made all other Exceptions redundant. I did not touch that old method. 
Also The parent method calling this just throws IOException in try catch, irrespective of the type of error. Check line 228 .

Comment (sahusanket) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java]: Refactored a bit and reverted the exceptions.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java [hunk 7]
```diff
@@ -392,61 +379,94 @@ private void importFromGcs(String projectId, TableReference tableRef, @Nullable
...
+
+        if (operation.equals(Operation.INSERT)) { // For the case when gcs paths is more than 10000
+          handleInsertOperation(tableRef, writeDisposition, loadConfig.getDestinationEncryptionConfiguration(),
+                                projectId, jobId, dataset, tableExists);
+        } else {
+          handleUpdateUpsertOperation(tableRef, kmsKeyName, getJobIdForUpdateUpsert(conf), conf);
+        }
+      }
+
+      setTemporaryTableExpiration();
```
Comment (fernst) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java:407]: I'd like for the temporary table to get created and the expiration set before the load in batches happens.

Think about it. If there are 2 batches, first succeeds and second fails, the table will not have an expiration set, thus is will stay as a dangling table.

Since I do not want to block the RC5 release, let's merge this and address this in a later PR.

**Review by fernst — APPROVED** [review]
Looks good, just left one comment.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java (hunk 7 above)
Comment (fernst) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java:407]: This can be achieved by creating the temp table with a TTL and then load records while allowing for the table schema to be relaxed

File: src/test/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormatTest.java [hunk 8]
```diff
@@ -0,0 +1,140 @@
+package io.cdap.plugin.gcp.bigquery.sink;
```
Comment (CuriousVini) [inline src/test/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormatTest.java:16]: Please include license  with year 2021. similar to 
https://github.com/data-integrations/google-cloud/blob/develop/src/main/java/io/cdap/plugin/gcp/bigquery/sink/AbstractBigQuerySink.java#L2

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java [hunk 9]
```diff
@@ -161,6 +162,8 @@ public OutputCommitter createCommitter(TaskAttemptContext context) throws IOExce
     private boolean allowSchemaRelaxation;
     private boolean allowSchemaRelaxationOnEmptyOutput;
 
+    private static final int BQ_IMPORT_MAX_BATCH_SIZE = 10000;
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java:165]: Should we document this limitation in bq sink documentation?

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java [hunk 10]
```diff
@@ -378,6 +356,15 @@ private void importFromGcs(String projectId, TableReference tableRef, @Nullable
             JobInfo.SchemaUpdateOption.ALLOW_FIELD_ADDITION.name(),
             JobInfo.SchemaUpdateOption.ALLOW_FIELD_RELAXATION.name()));
         }
+      } else {
+        if (!tableExists && Operation.UPSERT.equals(operation)) {
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java]: nit: move else if in same line similar to:
https://github.com/data-integrations/google-cloud/blob/f054621188c1301d7d47b204b2d4888f6a7a08d3/src/main/java/io/cdap/plugin/gcp/gcs/actions/GCSArgumentSetter.java#L157

Comment (CuriousVini) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java]: We dont need this check for insert?

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java [hunk 11]
```diff
@@ -392,61 +379,94 @@ private void importFromGcs(String projectId, TableReference tableRef, @Nullable
...
+
       Job job = new Job();
-      job.setConfiguration(config);
+      job.setConfiguration(jobConfiguration);
       job.setJobReference(jobReference);
-
-      // Insert and run job.
       bigQueryHelper.insertJobOrFetchDuplicate(projectId, job);
-
-      // Poll until job is complete.
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java:413]: why this change is needed?

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java [hunk 12]
```diff
@@ -392,61 +379,94 @@ private void importFromGcs(String projectId, TableReference tableRef, @Nullable
...
         new JobReference().setProjectId(projectId)
           .setJobId(jobId)
           .setLocation(dataset.getLocation());
+
       Job job = new Job();
-      job.setConfiguration(config);
+      job.setConfiguration(jobConfiguration);
       job.setJobReference(jobReference);
-
-      // Insert and run job.
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java:410]: why this change is needed?

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java (hunk 10 above)
Comment (sahusanket) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java]: For INSERT , since it's a Big Query Job directly into destination table, it Automatically creates the Table.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java (hunk 9 above)
Comment (sahusanket) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java:165]: @CuriousVini  this limitation is addressed now via creating temp table and loading and User would not have to take any extra action.  We can mention this temp table thing in doc But I went through the Docs in this repo, we do not have a section for any such internal processing notes that does not require user action.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java (hunk 12 above)
Comment (sahusanket) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java:410]: reverted

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java (hunk 11 above)
Comment (sahusanket) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java:413]: reverted

File: src/test/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormatTest.java (hunk 8 above)
Comment (sahusanket) [reply src/test/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormatTest.java:16]: added
